plutus-use-cases/src/Language/Plutus/Coordination/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,164 @@
+    , contribute
+    , contributionScript
+    , refund
+    , refundTrigger
+    -- * Functionality for campaign owners
+    , collect
+    , collectFundsTrigger
+    ) where
+
+import           Language.Plutus.Coordination.Plutus
+import           Language.Plutus.CoreToPLC.Plugin     (plc)
+import qualified Language.Plutus.CoreToPLC.Primitives as Prim
+
+import           Prelude                              (Bool (..), Either (..), Num (..), Ord (..), succ, sum, ($))
+
+-- | A crowdfunding campaign.
+data Campaign = Campaign
+    { campaignDeadline           :: !BlockHeight
+    , campaignTarget             :: !Value
+    , campaignCollectionDeadline :: !BlockHeight
+    , campaignOwner              :: !PubKey
+    }
+
+-- | Contribute funds to the campaign (contributor)
+--
+contribute :: Campaign -> Value -> TxM [TxOutRef]
+contribute c value = do
+    assert (value > 0)
+    contributorPubKey <- lookupMyPubKey
+    myPayment         <- createPayment (value + standardTxFee)
+    let validator = contributionScript c contributorPubKey
+        o = TxOutScript
+            value
+            (hash validator)
+            0 -- TODO: contributorPubKey ought to be lifted into PLC at coordination runtime as the data script
+    submitTransaction Tx
+      { txInputs  = Left myPayment -- TODO: Change to [myPayment] when we can have a list of inputs
+      , txOutputs = Left o
+      }
+    -- the transaction above really ought to be merely a transaction *template* and the transaction fee ought to be
+    -- added by the Wallet API Plutus library on the basis of the size and other costs of the transaction
+
+-- | The validator script that determines whether the campaign owner can
+--   retrieve the funds or the contributors can claim a refund.
+--
+--   Assume there is a campaign `c :: Campaign` with two contributors
+--   (identified by public key pc_1 and pc_2) and one campaign owner (pco).
+--   Each contributor creates a transaction, t_1 and t_2, whose outputs are
+--   locked by the scripts `contributionScript c pc_1` and `contributionScript
+--   c pc_1` respectively.
+--   There are two outcomes for the campaign.
+--   1. Campaign owner collects the funds from both contributors. In this case
+--      the owner creates a single transaction with two inputs, referring to
+--      t_1 and t_2. Each input contains the script `contributionScript c`
+--      specialised to a contributor. This case is covered by the
+--      definition for `payToOwner` below.
+--   2. Refund. In this case each contributor creates a transaction with a
+--      single input claiming back their part of the funds. This case is
+--      covered by the `refundable` branch.
+contributionScript ::
+       Campaign
+    -> PubKey
+    -> PlutusTx
+contributionScript _ _  = PlutusTx inner where
+
+    --   See note [Contracts and Validator Scripts] in
+    --       Language.Plutus.Coordination.Contracts
+    inner = plc  (\() () p Campaign{..} contribPubKey ->
```

Comment:
I notice that you're using `plc` directly rather than the TH function. I think we probably want to force people to go through the latter - but I guess GHC does better checking for you this way?

Reply:
I didn't notice any differences in the type checking, the main reason is that the TH notation causes problems with hlint and stylish-haskell. But I changed in CrowdFunding.hs, and left the `plc` notation in `Swap.hs` (although the plugin is currently deactivated there) so now we have examples of both.